File: server/src/game/Game.ts
// Game gom 2 Player + Turn — orchestration: switchTurn, getOpponent, isFinished.
import { Player } from './Player.js';
import { GameStatus } from '@coincard/shared';
import { FIRST_PLAYER_HAND_SIZE, SECOND_PLAYER_HAND_SIZE } from './constants.js';

export class Game {
  private status: GameStatus = GameStatus.WAITING;
  private turn = 0;
  private activePlayerId = '';
  private winnerId: string | null = null;
  private statusMessage = '';
  private manualDrawTurn = -1;

  constructor(
    public readonly gameId: string,
    public readonly roomCode: string,
    private readonly players: [Player, Player],
  ) {}

  start(): void {
    if (this.status !== GameStatus.WAITING) throw new Error('Trận đã bắt đầu.');
    const firstPlayer = this.players[0];
    const secondPlayer = this.players[1];
    if (!firstPlayer || !secondPlayer) {
      throw new Error('Game cần đúng 2 player.');
    }
    firstPlayer.drawAndAddToHand(FIRST_PLAYER_HAND_SIZE);
    secondPlayer.drawAndAddToHand(SECOND_PLAYER_HAND_SIZE);
    // Đảm bảo tay đầu luôn có nước đi (lá ≤ 2 mana).
    firstPlayer.guaranteeCheapOpener(2);
    secondPlayer.guaranteeCheapOpener(2);
    this.activePlayerId = firstPlayer.id;
    this.turn = 1;
    this.status = GameStatus.PLAYING;
    this.setStatusMessage('HERO 1 STARTS');
    this.beginTurnForPlayer(0, { skipDraw: true });
  }

  switchTurn(): void {
    this.activePlayerId = this.getOpponent().id;
    this.turn++;
    this.setStatusMessage('');
    this.beginTurn();
  }

  isFinished(): boolean {
    return this.status === GameStatus.FINISHED;
  }

  getStatus(): GameStatus {
    return this.status;
  }

  getCurrentTurn(): number {
    return this.turn;
  }

  getActivePlayerId(): string {
    return this.activePlayerId;
  }

  getWinnerId(): string | null {
    return this.winnerId;
  }

  hasManualDrawnThisTurn(): boolean {
    return this.manualDrawTurn === this.turn;
  }

  markManualDraw(): void {
    this.manualDrawTurn = this.turn;
  }

  getStatusMessage(): string {
    return this.statusMessage;
  }

  setStatusMessage(message: string): void {
    this.statusMessage = message;
  }

  getPlayers(): Readonly<Player[]> {
    return this.players.slice();
  }

  getOpponent(): Player {
    return this.players.find((p) => p.id !== this.activePlayerId) as Player;
  }

  getPlayerById(playerId: string): Player {
    const found = this.players.find((p) => p.id === playerId);
    if (!found) {
      throw new Error('Player id không tồn tại.');
    }
    return found;
  }

  resign(playerId: string): void {
    this.getPlayerById(playerId);
    if (this.isFinished()) {
      return;
    }
    const winner = this.players.find((p) => p.id !== playerId);
    this.finish(winner?.id ?? null, 'CONCEDE');
  }

  finish(winnerId: string | null, reason = ''): void {
    if (this.isFinished()) {
      return;
    }
    this.status = GameStatus.FINISHED;
    this.winnerId = winnerId;
    if (reason) {
      this.setStatusMessage(reason);
    }
  }

  private beginTurn(): void {
    this.beginTurnForPlayer(this.players.findIndex((p) => p.id === this.activePlayerId));
  }

  private beginTurnForPlayer(index: number, options: { skipDraw?: boolean } = {}): void {
    const player = this.players[index];
    if (!player) {
      throw new Error('Player index không hợp lệ.');
    }
    player.increaseMaxMana();
    player.refillMana();
    player.getBoard().forEach((m) => m.startTurn());

    // Lượt mở màn không draw — tay đầu đúng spec 3/4 lá.
    if (options.skipDraw) {
      return;
    }
    if (player.deckSize > 0) player.addToHand(player.drawCard());
  }
}

File: server-java/src/main/java/vn/coincard/server/game/Game.java
package vn.coincard.server.game;

import java.util.List;
import java.util.Objects;

/** Mutable state for one authoritative match. */
public class Game {
  private GameStatus status = GameStatus.WAITING;
  private int turn;
  private String activePlayerId = "";
  private String winnerId;
  private String statusMessage = "";

  Game(String gameId, String roomCode, Player p1, Player p2) {
    this.gameId = required(gameId, "gameId");
    this.roomCode = required(roomCode, "roomCode");
    Player first = Objects.requireNonNull(p1, "p1");
    Player second = Objects.requireNonNull(p2, "p2");
    if (first.id().equals(second.id())) {
      throw new IllegalArgumentException("Hai player phải có id khác nhau.");
    }
    this.players = List.of(first, second);
  }

  private final String gameId;
  private final String roomCode;
  private final List<Player> players;

  public String getGameId() { return gameId; }
  public String getRoomCode() { return roomCode; }
  /** Starts a match with the official opening hand, Coin, and first-turn draw. */
  void start(String firstPlayerId) {
    if (status != GameStatus.WAITING) throw new IllegalStateException("Trận đã bắt đầu.");
    Player first = getPlayerById(firstPlayerId);
    Player second = players.stream().filter(p -> !p.id().equals(firstPlayerId)).findFirst()
        .orElseThrow(() -> new IllegalStateException("Cần đủ hai người chơi."));
    first.drawAndAddToHand(Constants.FIRST_PLAYER_HAND_SIZE);
    second.drawAndAddToHand(Constants.SECOND_PLAYER_HAND_SIZE);
    second.addToHand(TokenCards.coinCard());
    activePlayerId = first.id();
    turn = 1;
    status = GameStatus.PLAYING;
    beginTurnForPlayer(players.indexOf(first));
  }

  public void switchTurn() {
    activePlayerId = getOpponent().id();
    turn++;
    setStatusMessage("");
    beginTurn();
  }

  public boolean isFinished() { return status == GameStatus.FINISHED; }
  public GameStatus status() { return status; }
  public int getCurrentTurn() { return turn; }
  public String getActivePlayerId() { return activePlayerId; }
  public String getWinnerId() { return winnerId; }
  public String getStatusMessage() { return statusMessage; }
  public List<Player> getPlayers() { return List.copyOf(players); }

  public Player getOpponent() {
    return players.stream().filter(p -> !p.id().equals(activePlayerId)).findFirst()
        .orElseThrow(() -> new IllegalStateException("Player index không hợp lệ."));
  }

  public Player getPlayerById(String playerId) {
    return players.stream().filter(p -> p.id().equals(playerId)).findFirst()
        .orElseThrow(() -> new IllegalArgumentException("Player id không tồn tại."));
  }

  void resign(String playerId) {
    getPlayerById(playerId);
    if (isFinished()) return;
    String winner = players.stream().filter(p -> !p.id().equals(playerId)).findFirst()
        .map(Player::id).orElse(null);
    finish(winner, "CONCEDE");
  }

  void finish(String winnerId, String reason) {
    if (isFinished()) return;
    status = GameStatus.FINISHED;
    this.winnerId = winnerId;
    if (reason != null && !reason.isEmpty()) setStatusMessage(reason);
  }

  record State(GameStatus status, int turn, String activePlayerId, String winnerId,
      String statusMessage, List<Player.State> players) {
    State {
      players = List.copyOf(players);
    }
  }

  State snapshotState() {
    return new State(status, turn, activePlayerId, winnerId, statusMessage,
        players.stream().map(Player::snapshotState).toList());
  }

  void restoreState(State state) {
    if (state.players().size() != players.size()) {
      throw new IllegalArgumentException("Sai số lượng player snapshot.");
    }
    for (int index = 0; index < players.size(); index++) {
      players.get(index).restoreState(state.players().get(index));
    }
    status = state.status();
    turn = state.turn();
    activePlayerId = state.activePlayerId();
    winnerId = state.winnerId();
    statusMessage = state.statusMessage();
  }

  private void beginTurn() {
    int index = 0;
    for (int i = 0; i < players.size(); i++) {
      if (players.get(i).id().equals(activePlayerId)) index = i;
    }
    beginTurnForPlayer(index);
  }

  private void beginTurnForPlayer(int index) {
    Player player = players.get(index);
    player.increaseMaxMana();
    player.refillMana();
    player.getBoard().forEach(Minion::startTurn);
    player.drawForTurn();
  }

  private void setStatusMessage(String message) { statusMessage = message; }

  private static String required(String value, String field) {
    if (value == null || value.isBlank()) throw new IllegalArgumentException(field + " là bắt buộc.");
    return value;
  }
}
